client/snap_op.go

```diff
@@ -261,21 +289,30 @@ func (client *Client) Try(path string, options *SnapOptions) (changeID string, e
 	return client.doAsync("POST", "/v2/snaps", nil, headers, buf)
 }
 
-func sendSnapFile(snapPath string, snapFile *os.File, pw *io.PipeWriter, mw *multipart.Writer, action *actionData) {
-	defer snapFile.Close()
+func sendSnapFiles(paths []string, files []*os.File, pw *io.PipeWriter, mw *multipart.Writer, action *actionData) {
+	defer func() {
+		for _, f := range files {
+			f.Close()
+		}
+	}()
```

Comment:
Would this be simpler if the responsibility for closing was moved to the caller? Moving this to `InstallPathMany` would get rid of the err handling case (`for _, openFile := range files`..) in `InstallPathMany`; you would then need to add an explicit `defer f.Close() in `InstallPath`.

Reply:
I think it should be fine, even if doAsync() fails it will close the reader end, thus making writes fail too. If you want, you can have additional sync with sendSnapFiles exiting in sendLocalSnaps like so:
```
waitChan := make(chan struct{})
go func() {
	sendSnapFiles(paths, files, pw, mw, &action)
	close(waitChan)
}
defer func() {
	<-waitChan
}
...
_, changeID, err := client.doAsyncFull("POST", "/v2/snaps", nil, headers, pr, doNoTimeoutAndRetry)
if err != nil {
	// make sure the writer exits
	pr.Close()
} 
return changeID, err
```

Reply:
Maybe I'm misunderstanding something but moving the deferred close to the InstallPath(Many) function gives me the following order of operations:
```
calling doAsync
exiting doAsync
closing file (deferred)
closing file (deferred)
writing file in sendSnapFiles
```
So there would have to be some extra synchronization between the goroutines ensuring we only close the files after the writes. Leaving the closing of the files in the writer goroutine seems simpler. But maybe I missed the point